rust_snuba/rust_arroyo/src/processing/strategies/reduce.rs

```diff
@@ -0,0 +1,285 @@
+use crate::processing::strategies::{
+    merge_commit_request, CommitRequest, InvalidMessage, MessageRejected, ProcessingStrategy,
+    SubmitError,
+};
+use crate::types::{AnyMessage, InnerMessage, Message, Partition};
+use std::collections::BTreeMap;
+use std::mem;
+use std::sync::Arc;
+use std::time::{Duration, Instant, SystemTime};
+
+struct BatchState<T, TResult> {
+    value: Option<TResult>,
+    accumulator: Arc<dyn Fn(TResult, T) -> TResult + Send + Sync>,
+    offsets: BTreeMap<Partition, u64>,
+    batch_start_time: SystemTime,
+    message_count: usize,
+}
+
+impl<T: Clone, TResult: Clone> BatchState<T, TResult> {
+    fn new(
+        initial_value: TResult,
+        accumulator: Arc<dyn Fn(TResult, T) -> TResult + Send + Sync>,
+    ) -> BatchState<T, TResult> {
+        BatchState {
+            value: Some(initial_value),
+            accumulator,
+            offsets: Default::default(),
+            batch_start_time: SystemTime::now(),
+            message_count: 0,
+        }
+    }
+
+    fn add(&mut self, message: Message<T>) {
+        let tmp = self.value.take();
+        self.value = Some((self.accumulator)(tmp.unwrap(), message.payload()));
```

Comment:
Tried this but I get the following:
```
error[E0507]: cannot move out of `self.value` which is behind a mutable reference
  --> rust_arroyo/src/processing/strategies/reduce.rs:35:41
   |
35 |         self.value = (self.accumulator)(self.value, message.payload());
   |                                         ^^^^^^^^^^ move occurs because `self.value` has type `TResult`, which does not implement the `Copy` trait
```